src/main/cpp/Auto/SwerveAutoPath.cpp

```diff
@@ -1,34 +1,121 @@
-#include "Drive/AutoPath.h"
+#include "Auto/SwerveAutoPath.h"
 
 #include "Drive/DriveConstants.h"
+#include "Auto/AutoConstants.h"
 
 #ifndef M_PI
 #define M_PI 3.141592653589793238462643383279502884197169399
 #endif
 
+#include <frc/smartdashboard/SmartDashboard.h>
+
 /**
  * Constructor
 */
-AutoPath::AutoPath() : 
+SwerveAutoPath::SwerveAutoPath(SwerveControl& drivebase, std::vector<AutoPaths::SwervePose> poses) : 
   m_curAng{0}, m_multiplier{0}, m_curAngVel{0}, m_prevTime{0}, m_prevTimeOdom{0},
   m_calcTrans{100}, m_calcAng{100}, m_curState{NOT_EXECUTING},
-  m_kPPos{0}, m_kIPos{0}, m_kDPos{0}, m_kPAng{0}, m_kIAng{0}, m_kDAng{0}
+  m_kPPos{0}, m_kIPos{0}, m_kDPos{0}, m_kPAng{0}, m_kIAng{0}, m_kDAng{0}, drivebase_{drivebase}
 {
   using namespace AutoConstants;
   SetPosPID(TRANS_KP, TRANS_KI, TRANS_KD);
   SetAngPID(ANG_KP, ANG_KI, ANG_KD);
+  calculateTotalDistance();
+  AddPoses(poses);
+
+  // shuffleboard init
+  frc::SmartDashboard::PutNumber("length of poses array recieved", poses.size());
+  frc::SmartDashboard::PutNumber("waypoints completed", waypoints_completed);
+
+  frc::SmartDashboard::PutNumber("swerve kp pos", m_kPPos);
+  frc::SmartDashboard::PutNumber("swerve ki pos", m_kIPos);
+  frc::SmartDashboard::PutNumber("swerve kd pos", m_kDPos);
+
+  frc::SmartDashboard::PutNumber("swerve kp ang", m_kPAng);
+  frc::SmartDashboard::PutNumber("swerve ki ang", m_kIAng);
+  frc::SmartDashboard::PutNumber("swerve kd ang", m_kDAng);
+  frc::SmartDashboard::PutNumber("Swerve Auto Completion", m_completion);
 }
 
+/**
+ * @brief Recalculates the total distance that will be travelled by this spline
+ */
+void SwerveAutoPath::calculateTotalDistance() {
+  total_distance = m_calcTrans.getMaxDistance(0.1);
+}
+
+/**
+ * @brief Calculates the current progress, by averaging the x, y, and angle components.
+ */
+void SwerveAutoPath::calculateCurrentProgress() {
+  double curr_x = m_curPos.x();
+  double curr_y = m_curPos.y();
+  double num_waypoints = m_calcTrans.getAllWaypoints().size();
+
+  SwerveAutoPath::Pose2 previous_waypoint = m_calcTrans.getAllWaypoints()[waypoints_completed > 0 ? waypoints_completed - 1 : 0];
+  SwerveAutoPath::Pose2 current_waypoint = m_calcTrans.getAllWaypoints()[waypoints_completed];
+  // double x_completion = curr_x / current_waypoint.getPos().at(0);
+  // double y_completion = curr_y / current_waypoint.getPos().at(1);
+  double x_completion = calcGenericCompletion(previous_waypoint.getPos().at(0), curr_x, current_waypoint.getPos().at(0));
+  double y_completion = calcGenericCompletion(previous_waypoint.getPos().at(1), curr_y, current_waypoint.getPos().at(1));
+  double averaged = (x_completion + y_completion) / 2.0;
+
+  if (averaged == 1.0) {
```

Comment:
bro's not reaching target ever

Reply:
use a tolerance

Reply:
I would also recommend using the time to figure out what segment you're on, instead of incrementing whenever  it reaches the target. (your choice)